smtpserver/server.go

```diff
@@ -1899,6 +1900,34 @@ func (c *conn) submit(ctx context.Context, recvHdrFor func(string) string, msgWr
 		xsmtpUserErrorf(smtp.C550MailboxUnavail, smtp.SePol7DeliveryUnauth1, "must match authenticated user")
 	}
 
+	// Check if the message contains non-ascii characters. If no such characters are found,
+	// the SMTPUTF8 extension is not required.
+	// ../rfc/6531:497
+	isASCII := func(s string) bool {
+		for _, c := range s {
+			if c > unicode.MaxASCII {
+				return false
+			}
+		}
+		return true
+	}
+	c.smtputf8 = !isASCII(c.mailFrom.Localpart.String())
```

Comment:
Localpart.String() marshals the Localpart string type to an encoded string. that won't change whether the string has non-ascii. so we can just check for characters with "isASCII(string(c.mailFrom.Localpart))".